platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+8. The roster must have a unique NodeId for each RosterEntry, and the NodeId key must match the NodeId value in the
+   corresponding RosterEntry.
+
+On the submission of a new Candidate Roster, the platform will validate the roster against these conditions.
+Note that a constructed `Roster` can be valid, but not accepted by the platform.
+For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing
+candidate roster, the new roster will be discarded. That is, the operation has no effect.
+
+## Data storage
+
+### State
+
+All the data related to Active and Candidate Roster will be stored in the State.
+
+Two new objects will be added to the state:
+
+1. A `RosterState` singleton object that will store the current candidate roster hash and a list of round numbers
+   and hashes of the active roster used to sign each round.
+2. A map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```. All candidate and active rosters will be
+   stored here.
+   When a new candidate roster is received, it will be stored in this map, with the `RosterState`'s current candidate
+   roster hash updated.
+
+### On Disk
+
+There will continue to be components used in creating the Roster - such as the private EC (elliptic curve) key, the RSA
+private key and the signing X509Certificate certificate - that will be stored on disk.
+However, it is at the discretion of the Services team to manage the lifecycle of these files, and not the platform.
+
+There will not be any other separate artifacts from this proposal stored directly on disk.
+
+## Services and DevOps changes (Inversion of control)
+
+The current startup procedure will be altered as follows
+
+- The `config.txt` file will no longer be used by the platform for storing the address book in existing networks. It
+  will no longer be used by the platform code
+- Designating `config.txt` introduces inversion of control. It will be at the exclusive prerogative of Services (the
+  App) going forward. Platform will no longer be responsible for it, and all the platform code that builds an
+  AddressBook from `config.txt` will be removed or refactored and moved into the Services codebase.
+
+The current network Transplant procedure is manual.
+DevOps get given a State and `config.txt` file on disk. This is then followed by a software-only upgrade to adopt
+the `config.txt` file using that state. This will change to the use of a Genesis Roster.
+
+#### Genesis Roster
+
+![](Proposed%20Network%20Transplant%20flow.drawio.svg)
+A `Genesis Roster` is an optional Roster that DevOps may provide to either start a Genesis Network process or Transplant
+an existing network. It is essential to distinguish the Genesis Roster from the Candidate or Active rosters.
+The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network. However, its structure is
+exactly the same as the Active or Candidate Rosters.
```

Comment:
There may be a difference between a "Genesis Roster" (which should only be used at Genesis) and an "Override Roster" (which should override whatever is already in state and override the Genesis Roster). We will need the same things for configs -- genesis, override, and what is in state.

Reply:
It makes sense to name them this way. I have gotten confused with the term "genesis" referring to something that is not related to the genesis of a network.

Reply:
@rbair23, In your mental model of `override roster` is the network starting over from round 1 or trying to continue from the current round?    Does the hashgraph continue seemlessly?   Or is this a disjoint hashgraph?    When the ledger id enters the picture, these overrides will force a reset of the ledger id.

Reply:
@edward-swirldslabs I think the value is clarity. I agree we should scope the "Genesis Roster" to the case where the network starts at Genesis (round 1 e.g. Genesis Network Process) and an "Override Roster" to the case of a network transplant (round number continues from the last in the state). I'll clarify this in the doc (straightforward diagram change).

Reply:
@edward-swirldslabs I haven't dug into what actions are taken for override. However, we clearly have two different flows -- one where we are at genesis (there is no pre-existing state) and one where we are not at genesis. If we have one override file that behaves differently based on context (i.e. is there a saved state or not), then it becomes exceptionally tricky to work with. The node operator *must* deploy the roster on genesis, and then *must* remove it between the time the network starts and saves a state and the next node restart.

If instead you have two files -- one for genesis, and one for override purposes, then under normal circumstances the node operator either: (a) deploys one for genesis because their node is a genesis node, and then leave that file there for as long as they like (forever even!), or (b) never deploys a file because they're joining an existing network.

But they never have to worry about removing a file between startup and restart. And the only time to use an override file is if there is a **major** problem, or you are moving state to a new network (from mainnet to testnet or something).

Reply:
I've modified the text and diagrams to detail the need and behaviour of the override roster.